File: C-Sharp/other/CountInversions.cs
//Count Inversion problem with mergesort algorithm where i < j and arr[j] < arr[i] is an inversion
using System;

namespace CountInversionsProblem
{
    class Count_Inversions
    {
        //The merge step to sort the elements and count inversions
        private static int Merge(int[] arr, int[] temp, int leftStart, int mid, int rightEnd)
        {
            int left = leftStart, right = mid + 1, index = left, count = 0;
            //Compare Elments from arr portion and sort them in temproray array temp
            while (left < mid + 1 && right <= rightEnd)
            {
                if (arr[right] < arr[left])
                {
                    temp[index] = arr[right++];
                    count += mid + 1 - left;
                }
                else
                {
                    temp[index] = arr[left++];
                }
                index++;
            }
            // if the arr porion sorted but i didn't reach the last element place the rest in temp 
            for (; left <= mid; left++)
                temp[index++] = arr[left];
            // if the arr porion sorted but j didn't reach the last element place the rest in temp 
            for (; right <= rightEnd; right++)
                temp[index++] = arr[right];
            //Place the sorted elements from temp into arr
            for (left = leftStart; left <= rightEnd; left++)
                arr[left] = temp[left];

            return count;
        }
        // recursive part in merge sort to break the array into smaller portions and count inversions
        private static int CountInversions(int[] arr, int[] temp, int leftStart, int rightEnd)
        {
            if (leftStart < rightEnd)
            {
                int count = 0;
                int mid = leftStart + (rightEnd - leftStart) / 2;
                count += CountInversions(arr, temp, leftStart, mid);
                count += CountInversions(arr, temp, mid + 1, rightEnd);
                count += Merge(arr, temp, leftStart, mid, rightEnd);
                return count;
            }
            return 0;

        }
        //Limit the number of parameters
        public static int CountInversions(int[] arr)
        {
            int count = CountInversions(arr, new int[arr.Length], 0, arr.Length - 1);
            return count;
        }
        static void Main(string[] args)
        {
            Console.Write("Enter array size: ");
            int n = Convert.ToInt32(Console.ReadLine());
            int[] arr = new int[n];
            Console.WriteLine("Enter array Elements: ");
            for (int i = 0; i < n; i++)
            {
                arr[i] = Convert.ToInt32(Console.ReadLine());
            }
            int count = CountInversions(arr);
            Console.WriteLine($"\n\nNumber of inversions = {count}");
        }
    }
}
/*
Sample Input


Enter array size: 5
Enter arry elements : 
2
1
3
1
2

input array : 2 1 3 1 2

number of inversions output : 4 
Complexity:
Merge sort takes time complexity of O(nlogn) and space complexity of O(n)
 
 */

File: C/sort/Count_Inversions.c
/*
Problem Statement:
Given an array of integers. Find the Inversion Count in the array.
Two elements a[i] and a[j] form an inversion if a[i] > a[j] and i < j.
*/

#include <stdio.h>

// answer stores the inversion count.
int answer = 0;

// Merge function helps us to calculate the inversion count.
void Merge(int input[], int start, int end, int mid)
{
    // temp is an array, which finally stores the elements of input array in sorted manner.
    int temp[end - start + 1];
    int left = start, right = mid + 1, new_ind = 0;
    while (left <= mid && right <= end)
    {
        if (input[left] <= input[right])
        {
            temp[new_ind++] = input[left++];
        }
        else
        {
            // In this case, inversions are possible.
            answer += (mid - left + 1);
            temp[new_ind++] = input[right++];
        }
    }
    while (left <= mid)
    {
        temp[new_ind++] = input[left++];
    }
    while (right <= end)
    {
        temp[new_ind++] = input[right++];
    }
    for (int index = start; index <= end; index++)
    {
        input[index] = temp[index - start];
    }
    return;
}

void MergeSort(int input[], int start, int end)
{
    if (start >= end)
    {
        return;
    }
    int mid = start + (end - start) / 2;
    MergeSort(input, start, mid);
    MergeSort(input, mid + 1, end);
    Merge(input, start, end, mid);
}

int main()
{
    printf("Enter the size of the input array\n");
    int SIZE;
    scanf("%d", &SIZE);
    int input[SIZE];
    printf("Enter the elements of the input array\n");
    for (int i = 0; i < SIZE; i++)
    {
        scanf("%d", &input[i]);
    }
    MergeSort(input, 0, SIZE - 1);
    printf("Number of inversions in the input array are %d.", answer);
}

/*
Time Complexity  : O(SIZE*log(SIZE)).
Space Complexity : O(SIZE).

Input:
Enter the size of the input array
5
Enter the elements of the input array
2 4 1 3 5
Output:
Number of inversions in the input array are 3.

*/

